### ISSUE TITLE
Curiosity about circuit result derivation plot

### ISSUE THREAD

**Issue by minminjao** [RitajitMajumdar/Optimizing-Ansatz-Design-in-QAOA-for-Max-cut#1]
Hello, may I know why the code you wrote appears in front of "Figure (1561.68x445.48)" when you print(qc.draw(output="mpl", fold=-1)? I'm implementing it on Jupiter laptop, so I tried installing qiskit visualization separately, but it still doesn't come out on mpl, can you tell me how to fix it?
@RitajitMajumdar 

And can i know how to make this graph result? Use matplotlib?
![image](https://github.com/RitajitMajumdar/Optimizing-Ansatz-Design-in-QAOA-for-Max-cut/assets/108252034/ca0ec136-7ca7-42e9-93d4-34d14d5b5f65)

**Comment by ritajitmajumdar1** [comment]
Hello, I am not sure about the first one. In the tutorial.ipynb, I didn't see any such situation where the code is appearing in front of the figure. However, in general, if you comment out some part of code before doing print(qc.draw(output="mpl", fold=-1)), then usually that code gets printed. So please check if that is the case for you.

Yes, the graph result was plotted using matplotlib.

**Comment by minminjao** [comment]
Hello, thank you for your reply. When I applied the code you told me about, "Figure --" that I told you about appears. This is my error?( Currently running on Jupiter notebook).
Also, may I ask what code you made matplotlib? (If it is difficult to tell me, may I know the reference, etc.)? I'm asking because I've been looking for it for a few days and I haven't found the right one.)

@RitajitMajumdar

**Comment by ritajitmajumdar1** [comment]
Hi, you cannot generate this figure directly from the codes here. The codes here provide the optimised QAOA circuit only. If you generate a graph, and feed it to this code, it will give you the corresponding optimised circuit. Now you need to execute those circuits on some backend (maybe noisy) and calculate the fidelity. The graph that you pointed out, from the paper, plots the fidelity. This execution, calculation of fidelity you need to do separately, it is not here in this codebase.

**Comment by minminjao** [comment]
Hello, I checked what you said. I can't make the error rate written in your paper by coding, so I applied Qiskit's "plot_error_map" to derive the error rate, so could you check if I applied it well? If not, I wonder if it is correct to try to review the error rate by applying "plot_error_map".(We are trying to apply it to each circuit, 'Edge coloring, dfs'

@RitajitMajumdar

**Comment by ritajitmajumdar1** [comment]
You can simply use fake backend to do the noisy simulation: https://qiskit.org/documentation/apidoc/providers_fake_provider.html
Once you have the circuit qc, make your backend as the FakeBackend. For example, if you want noise model of, say ibmq_kolkata, make backend = FakeKolkata().
Now run your circuit like: counts = backend.run(qc, shots=1000).result().get_counts()
Similarly you can have the ideal counts where your backend is AerSimulator. Now you can calculate the fidelity of these two counts.

**Comment by minminjao** [comment]
Hello, I tried something similar to what you told me. I still need to modify the code. The above picture is the code I'm writing and now I'm working on the error map part little by little. I fixed it based on the code you told me, and if you could give me feedback if it's correct once, I'd appreciate it. If you can't because you're busy, please review if I'm trying.

<I'm going to try it again with the code below>
#get a fake backend from the fake provider
backend = FakeManilaV2()

#create a simple circuit

==> maybe use qc

#transpile the ideal circuit to a circuit that can be directly executed by the backend
transpiled_circuit = transpile(qc, backend)
//transpiled_circuit.draw('mpl')  <== I think this code not necessary

#Run
job = backend.run(transpiled_circuit)
counts = job.result().get_counts()
plot_histogram(counts)

#plot_error_map
plot_error_map(backend)

**Comment by ritajitmajumdar1** [comment]
Yes, this code seems correct. I am not sure what you want to generate now. Do you want to generate the graph from my paper that you posted in your first comment? If yes, then your final portion is incorrect. That graph is not about the noise in the system, it is about the fidelity. So here is something you can do:

You have the count dictionary from FakeManila backend. I am calling it noisy_count. You also have the count dictionary from Qasm simulator. I am calling it ideal_count. Now you want to find the fidelity between the noisy and ideal count. I did this in a different way for my paper. But here is a simple way you can try. Just import hellinger_fidelity from qiskit.quantum_info, and find

hellinger_fidelity(ideal_count, noisy_count)

This now gives you a single point of the graph. Even for that point, you should do averaging over a few trials. Similarly, you change the number of qubits, do this whole step, and get other points.

**Comment by minminjao** [comment]
Hello, this is the result of output=mpl that I mentioned last time. I'm not sure if it's my fault. And I tried it based on what you left, and it was an error like this and I'm trying to fix it. Before that, is it true that what you said is to make a code like this? (Or can I proceed with one code?)

And thank you very much for your kind explanation and examples.

https://qiskit.org/documentation/stable/0.19/stubs/qiskit.quantum_info.hellinger_fidelity.html <= I was working on the code by referring to this document. 

![image](https://github.com/RitajitMajumdar/Optimizing-Ansatz-Design-in-QAOA-for-Max-cut/assets/108252034/58d68c30-8cae-4314-9b96-15c362a92a8b)
![image](https://github.com/RitajitMajumdar/Optimizing-Ansatz-Design-in-QAOA-for-Max-cut/assets/108252034/4904a1fc-45c8-4a8d-9004-14330a8febef)

@ritajitmajumdar1  @RitajitMajumdar

**Comment by ritajitmajumdar1** [comment]
If you look into the error, it says you are running a 7 qubit circuit on a 5 qubit device. If you were taking FakeManila, that is a 5-qubit device. So you can run circuits using upto 5 qubits.

**Comment by minminjao** [comment]
Hello, I modified the code as you advised, and I succeeded and got the result. So I have a question among the things you answered and I want to make sure that I understand it correctly.

1. So the graphs at the end of your paper represent how much noise is reduced? (Circuits, etc.)
![image](https://github.com/RitajitMajumdar/Optimizing-Ansatz-Design-in-QAOA-for-Max-cut/assets/108252034/9a0f3db6-6960-49cd-98e0-dc9bafcb73b1)

2. If question number 1 is correct, I applied it to the non-optimizing version, edge coloring version, and DFS version through "hellinger_fidelity" and when I got the result, DFS had the lowest fidelity value(?). So does this mean that the result is similar to the graph you're representing? (And can I understand that this means that the number of CNOTs is reduced a lot and at the same time, the error rate in the circuit is the lowest?)
<first image: no optimization, second: Edge coloring, third: DFS  // The red graph is a graph that I made randomly, and I haven't figured out what black represents noise and what red represents, and I think it's probably a wrong graph..>

![image](https://github.com/RitajitMajumdar/Optimizing-Ansatz-Design-in-QAOA-for-Max-cut/assets/108252034/703b05d5-b084-491f-949d-5742a161b4b4)

![image](https://github.com/RitajitMajumdar/Optimizing-Ansatz-Design-in-QAOA-for-Max-cut/assets/108252034/a6afe622-33c2-4c24-b422-34daa67fb7ad)

![image](https://github.com/RitajitMajumdar/Optimizing-Ansatz-Design-in-QAOA-for-Max-cut/assets/108252034/36453745-70f3-4730-80a1-a97eeaef3678)

@RitajitMajumdar @ritajitmajumdar1

**Comment by ritajitmajumdar1** [comment]
No, typically my graph is infidelity, i.e., 1-fidelity. So you should be getting that DFS has the best fidelity. It's odd you are getting the opposite. Anyway, when I did the code, I had the undo_gates option set to True. This creates an inverse of the circuit, so in the absence of noise you should be getting all zero. Now, I checked the fidelity with the all zero state. Anyway, these two approaches should almost be identical. But definitely try a few more things, e.g., I don't know what graph you are using - I used Erdos Renyi and for each point it was an average over 50 instances. Averaging is necessary.

As for CNOT count, you can actually print the number of CNOT operators and check that it has been lowered.

**Comment by minminjao** [comment]
So you're saying I'm wrong? It has been confirmed that the CNOT and Depth lengths are reduced. Based on your graph picture attached above, I thought it was similar to what I derived because DFS is at the top of the list intuitively. If you don't mind, can you check if I'm doing the code correctly? I'm trying to organize it based on the tutorial on your GitHub. 

In summary, I understand that the values of "hellinger_fiedlity" in the way you told me are expressed similar to the graph above.

Below, I uploaded the code I wrote to pdf.
[Qaoa7.pdf](https://github.com/RitajitMajumdar/Optimizing-Ansatz-Design-in-QAOA-for-Max-cut/files/12450079/Qaoa7.pdf)
[tutorial.pdf](https://github.com/RitajitMajumdar/Optimizing-Ansatz-Design-in-QAOA-for-Max-cut/files/12450080/tutorial.pdf)

@ritajitmajumdar1 @RitajitMajumdar

**Comment by ritajitmajumdar1** [comment]
You are not wrong. Well if you see you have obtained a fidelity of 0.9267 with edge colouring method, and 0.9217 with DFS. So in this case, edge colouring is slightly better than DFS. In my plot, graphs denote 1-fidelity. So if you do that DFS will be higher than edge colouring. But this can happen. Your values are almost the same. So it is not very bad. To get that plot like I have, you need averaging. First, you need to average over different start vertex of DFS.  You have already used something like DFS(G,0.7,1). Here 1 implies that DFS is started from vertex 1. You should be changing this value, and take average over all. And then you should also take average over different graph instances.

**Comment by minminjao** [comment]
Hello, I understand what you are saying. So, do you mean that in order to get more efficient results for DFS as the content of the paper, we should set the value of the current 'DFS(G,0.77,1)' code to 'DFS(G,0.7,1)'?

@ritajitmajumdar1

**Comment by ritajitmajumdar1** [comment]
No, it is fine whether you use 0.7 or 0.77 as your probability of edge. I assume you are using Erdos Renyi graphs. The value 1 implies that the DFS tree starts from vertex 1. But the tree could have started from 2, 3 etc. So you should change this value 1. Suppose you are using a 5vertex graph. Then you should do DFS(G,0.7,1), DFS(G,0.7,2), DFS(G,0.7,3), DFS(G,0.7,4), DFS(G,0.7,5); find fidelity for each and average over that. Then you should repeat this same for different G.

**Comment by minminjao** [comment]
Hello. I tried to modify the code on the DFS side as you advised me, and found out that the values specified in Edge coloring and DFS were different, and when I modified it, DFS came out better. For example, when I changed it to 'Edgecoloring(G,0.7)' and set it to 'DFS(G,0.7,1)', the 'hellinger_fidelity' of DFS was a higher number. If we get a higher number on the DFS side, can we say that we have reached a similar conclusion to the graphs on the last page of your paper?
I'm sorry to ask you a lot of questions even though you must be busy.
@ritajitmajumdar1

**Comment by ritajitmajumdar1** [comment]
Yes higher value for hellinger fidelity for DFS implies you have reached a similar conclusion as the graph on my paper.

**Comment by minminjao** [comment]
Thank you very much. Without you, I wouldn't have completed the code and understood it. Thank you very much for your help.

@ritajitmajumdar1
